ethcore/src/engines/validator_set/contract.rs

```diff
@@ -22,50 +22,48 @@ use ethereum_types::{H256, Address};
 use parking_lot::RwLock;
 use bytes::Bytes;
 
-use futures::Future;
-use native_contracts::ValidatorReport as Provider;
-
 use client::EngineClient;
 use header::{Header, BlockNumber};
 use machine::{AuxiliaryData, Call, EthereumMachine};
 
 use super::{ValidatorSet, SimpleList, SystemCall};
 use super::safe_contract::ValidatorSafeContract;
 
+use_contract!(validator_report, "ValidatorReport", "res/contracts/validator_report.json");
+
 /// A validator contract with reporting.
 pub struct ValidatorContract {
+	contract_address: Address,
 	validators: ValidatorSafeContract,
-	provider: Provider,
+	provider: validator_report::ValidatorReport,
 	client: RwLock<Option<Weak<EngineClient>>>, // TODO [keorn]: remove
 }
 
 impl ValidatorContract {
 	pub fn new(contract_address: Address) -> Self {
 		ValidatorContract {
+			contract_address,
 			validators: ValidatorSafeContract::new(contract_address),
-			provider: Provider::new(contract_address),
+			provider: validator_report::ValidatorReport::default(),
 			client: RwLock::new(None),
 		}
 	}
 }
 
 impl ValidatorContract {
-	// could be `impl Trait`.
-	// note: dispatches transactions to network as well as execute.
-	// TODO [keorn]: Make more general.
-	fn transact(&self) -> Box<Fn(Address, Bytes) -> Result<Bytes, String>> {
-		let client = self.client.read().clone();
-		Box::new(move |a, d| client.as_ref()
+	fn transact(&self, data: Bytes) -> Result<(), String> {
+		let client = self.client.read().clone().as_ref()
```

Comment:
You sure, you need both `clone()` and `.as_ref()`?

Reply:
you are right ;)

Reply:
definitely not, that's just sloppy refactoring :p